Claymore managed library

Kryptos-FR opened the pull request:
Depends on https://github.com/fragcolor-xyz/claymore/pull/1.

Kryptos-FR committed:
Common: add claymore project and bump version

Reorganize projects shared properties and libraries

Kryptos-FR committed:
Common: CBVar - add support for bytes array

Kryptos-FR committed:
Claymore: add unmanaged structs

Kryptos-FR committed:
Common: add Node wrapper

Kryptos-FR committed:
Claymore: add Claymore main class

Kryptos-FR committed:
Claymore: coverage tests

Kryptos-FR committed:
CI: build claymore lib and update coverage test

Kryptos-FR committed:
Bump samples

sinkingsugar commented on src/Fragcolor.Chainblocks.Claymore/Claymore.cs:
```diff
@@ -0,0 +1,125 @@
+      }
+    }
+
+    public static async Task<Variable> RequestDataAsync(string hash)
```
👏

Kryptos-FR replied:
It has a big caveat though. It only works if it is fine to call `cmlrPoll()` from any thread following the one that initially created the request.

That's because, in .NET's `async await` pattern, the continuation after each `await` can be executed on any of the thread pool's thread. There are ways to force the same thread to resume (using some custom synchronization context) but it's hard to do right and it might not work well with the treading model of the consuming library/application.

So if we can guarantee that, it's ok. If not, then maybe it's better to only keep the blocking (synchronous) version.

Kryptos-FR replied:
It seems to be working correctly now. That's something we can rethink about if we implement some kind of queue that can batch queries together.

sinkingsugar replied:
chainblocks can work from any thread too though.

sinkingsugar replied:
I suspect the issue if any is in the glue code/rust 🤔 
Chainblocks works with stackful coroutines so the "thread" or stack is always correct.

Kryptos-FR committed:
Claymore: wrap the request logic into a class

That way, ticking can be controlled by the consuming code.
Also supports cancellation.

Kryptos-FR committed:
Common: use IntPtr for return value of unmanaged delegate

Kryptos-FR committed:
Claymore: prefix native structs with Cl

Kryptos-FR committed:
Claymore: factorize request in base class

Kryptos-FR committed:
Claymore: add upload request

Kryptos-FR committed:
Fix CI build following changes in chainblocks repo

After merging branch 'wgpu-stable' (commit 1c0f8277859a07f9c90e165ca9f316eb73affb1c)
